### DISCUSSION TITLE
Population history

### DISCUSSION THREAD

**Discussion by juan11iguel** [esa/pygmo2#178] [Q&A] [answered]
I wonder if there is a way to store the population states for every generation i.e. to have a _population history_. This would allow to visualize how the population evolves over the algorithm generations/iterations.

I was thinking about a hacky way where my fitness function could update a file with the given decision vector and its fitness, but this would impact performance of the fitness function (not that it would matter much since my fitness function is already slow anyway.. 😅). Yet it would not have a way to associate each decision vector to some individual ID, or to which generation/iteration the decision vector belongs.

Just for reference, looking at how other optimization libraries handle this, [Evox](https://github.com/EMI-Group/evox) implementation can be a good reference with their implementation of an [Evaluation Monitor](https://evox.readthedocs.io/en/latest/api/monitors/eval.html) (equivalent to `algorithm.get_log`'?) and [Population Monitor](https://evox.readthedocs.io/en/latest/api/monitors/pop.html) objects. 

Related to [issue 158](https://github.com/esa/pygmo2/issues/158) since I guess this would have to be implemented through callbacks.

**Comment by bluescarni** [comment, marked as answer]
Hi @juan11iguel !

UDPs in pygmo are stateful objects, so if you want to track the history of the decision vectors "seen" by the problem you can just add a tiny bit of code to the objective function that appends the decision vector/fitness to a list to be stored as a data member of the UDP.

After your UDP has been wrapped in a pygmo problem, you can fetch a reference to it via the ``extract()`` method, and via the reference you can access the list stored as data member.

https://esa.github.io/pygmo2/problem.html#pygmo.problem.extract
